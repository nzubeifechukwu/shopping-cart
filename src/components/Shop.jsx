import { useEffect, useState } from "react";

export default function Shop() {
  const [products, setProducts] = useState([]);
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(true);
  const [cart, setCart] = useState([]);
  const [quantity, setQuantity] = useState(0);
  const [changed, setChanged] = useState(false);

  useEffect(() => {
    fetch("https://fakestoreapi.com/products")
      .then((response) => {
        if (response.status >= 400) {
          throw new Error("server error");
        }
        return response.json();
      })
      .then((data) => setProducts(data))
      .catch((error) => setError(error))
      .finally(() => setLoading(false));
  }, []);

  function addToCart(productId) {
    // Add product only if input quantity was changed
    if (changed) {
      const cartProduct = products.find((product) => product.id === productId);
      cartProduct.quantity = quantity;
      setCart(Array.from(new Set([...cart, cartProduct]))); // Don't repeat products
    }
    console.log(cart);
    setChanged(false);
  }

  return error ? (
    <h1>Oops! There's a network error!</h1>
  ) : loading ? (
    <h1>Loading...</h1>
  ) : (
    <section>
      {products.map((product) => (
        <article key={product.id}>
          <img src={product.image} alt={product.title} />
          <h2>{product.title}</h2>
          <p>${product.price}</p>
          <p>{product.description}</p>
          <form action="">
            <label htmlFor="quantity">Quantity </label>
            <input
              type="number"
              id="quantity"
              name="quantity"
              min={1}
              onChange={(e) => {
                setChanged(true);
                setQuantity(parseInt(e.target.value));
              }}
            />
          </form>
          <h2>{product.id}</h2>
          <button type="button" onClick={() => addToCart(product.id)}>
            Add to Cart
          </button>
        </article>
      ))}
    </section>
  );
}
