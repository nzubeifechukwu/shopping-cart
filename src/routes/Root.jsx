import { useState } from "react";
import { Outlet } from "react-router";
import { Link } from "react-router";

export default function Root() {
  const [products, setProducts] = useState([]);
  const [cart, setCart] = useState([]);
  // const [error, setError] = useState(null);

  return (
    <>
      <nav>
        <Link to="/">Home</Link>
        <Link to="shop">Shop</Link>
        <Link to="cart">Cart</Link>
      </nav>
      {/* <h1>Welcome home!</h1> */}
      <div className="mainBody">
        <Outlet context={[products, setProducts, cart, setCart]} />
      </div>
    </>
  );
}
