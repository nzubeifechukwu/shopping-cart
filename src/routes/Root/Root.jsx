import { useState } from "react";
import { Link, Outlet, useLocation } from "react-router";

import styles from "./root.module.css";

export default function Root() {
  const [cart, setCart] = useState([]);
  const location = useLocation();
  const totalCartItems = cart.reduce((prev, curr) => prev + curr.quantity, 0);

  return (
    <>
      <nav className={styles.nav}>
        <Link to="/">Home</Link>
        <Link to="shop">Shop</Link>
        <Link to="cart">
          Cart <span>{totalCartItems ? totalCartItems : ""}</span>
        </Link>
      </nav>
      <section className={styles.hero}>
        <h1>Zast E-Commerce Store</h1>
        {location.pathname === "/" && (
          <p>
            A one-stop online shop for all your needs. Head over to{" "}
            <Link to="shop">Shop</Link> for a list of our products and to fill
            your <Link to="cart">cart</Link>.
          </p>
        )}
      </section>
      <section className={styles.mainBody}>
        <Outlet context={[cart, setCart]} />
      </section>
    </>
  );
}
