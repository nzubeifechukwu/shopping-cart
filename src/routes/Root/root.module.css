nav {
  display: flex;
  justify-content: space-evenly;
  font-weight: 600;
  padding: 32px;
  background-color: #bc6c25;
  position: fixed;
  width: 100%;
}

nav > a {
  text-decoration: none;
  color: #fefae0;
  text-transform: uppercase;
}

nav > a:hover {
  color: #faedcd;
}

nav > a > span {
  color: #bc6c25;
  background-color: #fefae0;
  border-radius: 50%;
}

.hero {
  padding-top: 120px;
  background-color: #faedcd;
  padding-bottom: 40px;
  margin-bottom: 40px;
}

.hero > h1 {
  margin-bottom: 10px;
  margin-left: 40px;
  margin-right: 40px;
  text-transform: uppercase;
}

.footer {
  padding-bottom: 40px;
}
