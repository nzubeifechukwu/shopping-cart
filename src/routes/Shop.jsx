import { useEffect, useState } from "react";
import { useOutletContext } from "react-router";

export default function Shop() {
  const [products, setProducts] = useState([]);
  const [cart, setCart] = useOutletContext();
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(true);
  const [quantity, setQuantity] = useState(0);
  const [changed, setChanged] = useState(false);

  useEffect(() => {
    fetch("https://fakestoreapi.com/products")
      .then((response) => {
        if (response.status >= 400) {
          throw new Error("server error");
        }
        return response.json();
      })
      .then((data) => setProducts(data))
      .catch((error) => setError(error))
      .finally(() => setLoading(false));
  }, []);

  function addToCart(productId) {
    // Add product only if input quantity was changed
    if (changed) {
      const cartProduct = products.find((product) => product.id === productId);
      const newCart = [...cart];
      newCart.forEach((item, index) => {
        if (cartProduct.id === item.id) {
          newCart.splice(index, 1, cartProduct); // Don't repeat products
        }
      });
      cartProduct.quantity = quantity;
      setCart(Array.from(new Set([...newCart, cartProduct]))); // Don't repeat products
    }
    // console.log(cart);
    setChanged(false);
  }

  return error ? (
    <p>Oops! There's a network error!</p>
  ) : loading ? (
    <p>Loading...</p>
  ) : (
    <section>
      <h2>Items in our store...</h2>
      {products.map((product) => (
        <article key={product.id}>
          <h3>{product.title}</h3>
          <img src={product.image} alt={product.title} />
          <p>${product.price}</p>
          <p>{product.description}</p>
          <form action="">
            <label htmlFor="quantity">Quantity </label>
            <input
              type="number"
              id="quantity"
              name="quantity"
              min={1}
              onChange={(e) => {
                setChanged(true);
                setQuantity(parseInt(e.target.value));
              }}
            />
          </form>
          <button type="button" onClick={() => addToCart(product.id)}>
            Add to Cart
          </button>
        </article>
      ))}
    </section>
  );
}
