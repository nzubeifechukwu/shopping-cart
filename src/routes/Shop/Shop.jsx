import { useEffect, useState } from "react";
import { useOutletContext } from "react-router";

import styles from "./shop.module.css";

export default function Shop() {
  const [products, setProducts] = useState([]);
  const [cart, setCart] = useOutletContext();
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(true);
  const [quantity, setQuantity] = useState(0);
  const [changed, setChanged] = useState(false);

  useEffect(() => {
    fetch("https://fakestoreapi.com/products")
      .then((response) => {
        if (response.status >= 400) {
          throw new Error("server error");
        }
        return response.json();
      })
      .then((data) => setProducts(data))
      .catch((error) => setError(error))
      .finally(() => setLoading(false));
  }, []);

  function addToCart(productId) {
    // Add product only if input quantity was changed
    if (changed) {
      const cartProduct = products.find((product) => product.id === productId);
      const newCart = [...cart];
      newCart.forEach((item, index) => {
        if (cartProduct.id === item.id) {
          newCart.splice(index, 1, cartProduct); // Don't repeat products
        }
      });
      cartProduct.quantity = quantity;
      setCart(Array.from(new Set([...newCart, cartProduct]))); // Don't repeat products
    }
    setChanged(false);
  }

  return error ? (
    <p>Oops! There's a network error!</p>
  ) : loading ? (
    <p>Loading...</p>
  ) : (
    <>
      <h2>Items in our store</h2>
      <section className={styles.products}>
        {products.map((product) => (
          <article key={product.id} className={styles.article}>
            <h3>{product.title}</h3>
            <img src={product.image} alt={product.title} width={200} />
            <div>
              <p>${product.price}</p>
              <form action="">
                <label htmlFor="quantity">Qty</label>
                <input
                  type="number"
                  id="quantity"
                  name="quantity"
                  min={1}
                  onChange={(e) => {
                    setChanged(true);
                    setQuantity(parseInt(e.target.value));
                  }}
                />
              </form>
            </div>
            <button type="button" onClick={() => addToCart(product.id)}>
              Add to Cart
            </button>
          </article>
        ))}
      </section>
    </>
  );
}
