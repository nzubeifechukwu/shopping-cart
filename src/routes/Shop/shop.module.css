.initialP {
  margin-bottom: 60px;
  margin-left: 40px;
  margin-right: 40px;
}

.products {
  display: grid;
  gap: 20px;
  grid-template-columns: repeat(auto-fit, minmax(300px, 1fr));
  margin: 2rem 60px 60px;
}

.products > article {
  border-radius: 20px;
  padding: 1.5rem;
  background-color: #dda15e;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
}

.products > article > h3 {
  margin-bottom: 20px;
}

.products > article > div {
  display: flex;
  margin: 20px;
  font-weight: 500;
  align-items:flex-end;
}

.products > article > div > p {
  margin-right: 60px;
}

.products > article > div > form {
    display: flex;
    align-items: flex-end;
}

.products > article > div > form > label {
  margin-right: 5px;
  text-transform: uppercase;
}

.products > article > div > form > input {
  color: #bc6c25;
  font-weight: 600;
  width: 40px;
  padding: 4px;
  border: none;
  border-radius: 3px;
}

.products > article > button {
  background-color: #283618;
  color: #fefae0;
  padding: 15px;
  font-weight: 600;
  text-transform: uppercase;
  border-radius: 10px;
  border: none;
  cursor: pointer;
}

.products > article > button:hover {
  background-color: #606c38;
}
