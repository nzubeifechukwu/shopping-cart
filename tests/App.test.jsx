import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MemoryRouter, Outlet, Routes, Route } from "react-router";
import Root from "../src/routes/Root";
import Shop from "../src/routes/Shop";

// Test suite for Root component
describe("Root component", () => {
  it("renders component", () => {
    render(
      <MemoryRouter>
        <Root />
      </MemoryRouter>
    );

    screen.debug();
  });

  it("renders heading", () => {
    render(
      <MemoryRouter>
        <Root />
      </MemoryRouter>
    );

    expect(screen.getByRole("heading")).toBeInTheDocument();
  });

  it("renders nav", () => {
    render(
      <MemoryRouter>
        <Root />
      </MemoryRouter>
    );
    expect(screen.getByRole("navigation")).toBeInTheDocument();
  });

  it("renders links", () => {
    render(
      <MemoryRouter>
        <Root />
      </MemoryRouter>
    );
    expect(screen.getAllByRole("link")).toHaveLength(5);
  });

  it("renders text", () => {
    render(
      <MemoryRouter>
        <Root />
      </MemoryRouter>
    );
    expect(screen.getByText(/one-stop online shop/i)).toBeInTheDocument();
  });

  it("renders Home page when no nav link has been clicked", () => {
    // screen has no container property, so destructure render to get container
    const { container } = render(
      <MemoryRouter>
        <Root />
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders Shop page after Shop nav link is clicked", async () => {
    const user = userEvent.setup();
    render(
      <MemoryRouter>
        <Root />
      </MemoryRouter>
    );
    const navLinks = screen.getAllByRole("link", { name: "Shop" });
    await user.click(navLinks[0]);
    expect(await screen.findAllByRole("button")).toHaveLength(20);
  });
});

// Test suite for Shop component
// Mock the parent element
// const ParentComponentWithOutlet = ({ children }) => {
//   const outletContextValue = {
//     message: "Hello from Outlet Context",
//     count: 5,
//   };
//   return <Outlet context={outletContextValue} />;
// };
// describe("Shop component", () => {
//   it("renders the component with outlet context provided", () => {
//     render(
//       <MemoryRouter>
//         <Routes>
//           <Route path="/" element={<ParentComponentWithOutlet />}>
//             <Route element={<Shop />} />
//           </Route>
//         </Routes>
//       </MemoryRouter>
//     );
//     screen.debug();
//   });

//   it("renders data from the API call", async () => {
//     render(
//       <MemoryRouter>
//         <Routes>
//           <Route path="/" element={<ParentComponentWithOutlet />}>
//             <Route element={<Shop />} />
//           </Route>
//         </Routes>
//       </MemoryRouter>
//     );

//     expect(await screen.findAllByRole("button")).toHaveLength(20)

//     // expect(
//     //   await screen.findByText(/Your perfect pack for everyday use/)
//     // ).toBeInTheDocument();
//   });
// });
