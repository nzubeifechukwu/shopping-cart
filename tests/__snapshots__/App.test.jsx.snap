// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Root component > renders Home page when no nav link has been clicked 1`] = `
<div>
  <nav>
    <a
      data-discover="true"
      href="/"
    >
      Home
    </a>
    <a
      data-discover="true"
      href="/shop"
    >
      Shop
    </a>
    <a
      data-discover="true"
      href="/cart"
    >
      Cart
    </a>
  </nav>
  <section
    class="hero"
  >
    <h1>
      Zast E-Commerce Store
    </h1>
    <p>
      A one-stop online shop for all your needs. Head over to
       
      <a
        data-discover="true"
        href="/shop"
      >
        Shop
      </a>
       for a list of our products and to fill your 
      <a
        data-discover="true"
        href="/cart"
      >
        cart
      </a>
      .
    </p>
  </section>
  <div
    class="mainBody"
  />
</div>
`;
